### ISSUE TITLE
[Question] UndoManager.canUndo === true after create.

### ISSUE THREAD

**Issue by chemitaxis** [mobxjs/mobx-state-tree#524]
Hi guys, sorry for all my questions in the last weeks... 😭  

We are now creating our architecture for our editor (we are migrating from jQuery to React + MST, and we are preparing a series of blog posts to explain how the amazing and easy is being this process).

We are now writing our undo manager system and we want to create something very cool and powerful with MST. We want to connect our "items" models with our API. It can be very easy and reduce a lot of effort because MST supports the JSON Patch standard (cool again guys). 

But I think we have found a "bug" or "something strange" in the UndoManager middleware. This is our concept system with MST, very easy:

```
import { types, onPatch, getSnapshot } from "mobx-state-tree";
import { UndoManager } from "mst-middlewares";

// This is our "item" Todo 
const Todo = types
  .model("Todo", {
    id: types.identifier(),
    title: types.string,
    done: false
  })
  .actions(self => ({
    afterCreate() {
      onPatch(self, snapshot => {
       // If a patch is done in our item, we call to the API and update the data in our database
        console.log(`Request PATCH to API /todo/${self.id} with this body:`);
        console.log(snapshot);
      });
    },
    toggle() {
      self.done = !self.done;
    }
  }));

// Main Store:
const Store = types
  .model("Store", {
    todos: types.optional(types.map(Todo), {}),
    history: types.optional(UndoManager, {}),
    selectedTodos: types.optional(types.array(types.reference(Todo)), [])
  })
  .actions(self => ({
    createTodo(todo) {
      self.todos.put(todo);
    },
    selectTodo(todo) {
      self.selectedTodos.push(todo);
    },
    unSelectTodo(todo) {
      self.selectedTodos.remove(todo);
    },
    markAsDoneSelectedTodos() {
      self.selectedTodos.map(todo => todo.toggle());
    }
  }));

// create an instance from a snapshot
const store = Store.create({});

// ****************************************
// Code Question
// ****************************************
console.log(store.history.canUndo); // output: true, why? I just created the store...

console.log(store.history.history);

// Todo items:
const todo1 = Todo.create({ title: "Hello1", done: false, id: "1" });
const todo2 = Todo.create({ title: "Hello2", done: false, id: "2" });
const todo3 = Todo.create({ title: "Hello2", done: false, id: "3" });


// We don't want to attach this actions to the UndoManager, ok.
store.history.withoutUndo(() => {
  // Insert todo items in the store
  store.createTodo(todo1);
  store.createTodo(todo2);
  store.createTodo(todo3);
  // Select todo items
  store.selectTodo(todo3);
  store.selectTodo(todo2);
});

const snapShotBefore = getSnapshot(store);

// Mark as done, and yes, attach it to the UndoManager
store.markAsDoneSelectedTodos();
store.history.undo();

const snapShotAfter = getSnapshot(store);

// Check snapshots are identical before and after undo...
console.log(JSON.stringify(snapShotAfter.todos) === JSON.stringify(snapShotBefore.todos));
console.log(JSON.stringify(snapShotAfter.selectedTodos) === JSON.stringify(snapShotBefore.selectedTodos));
```

I have two questions... 

1.- in this line `console.log(store.history.canUndo);` Why is true the `canUndo`?? The user can't undo nothing... and I have just created the store. If I print the output of the history history prop this is the result:

```
▶[Object]
   0: Object
        patches: Array[0]
        inversePatches: Array[0]
```

2.- Second question and totally optional 😄  What do you think about this pattern? Do I need to know something important or are we making some concept mistake? 

I have created a codesandbox too, if you want to test it:

https://codesandbox.io/s/0y6q7vjzzp

Thank you so much for your amazing work!

**Comment by robinfehr** [comment]
I assume it's from your afterCreate action. you'd have to wrap that in a withoutUndo but trying that, you'll run into further problems :) there is work in progress that will fix this :)

**Comment by chemitaxis** [comment]
Thanks you so much for your answer @robinfehr ... but I can’t understand very well some concepts...

1. In this point you are talking about afterCreate method, are you talking about the afterCreation Todo model or about afterCreate Store model? I don’f fire the afterCreate “manually” in the Store model, is fired automatically?

About the workaround... Can you explain better please?

When you are talking about the fix progress... are you talking about this PR? https://github.com/mobxjs/mobx-state-tree/pull/504

2. What the benefits about create a view for selectedTodos? Do I need to reference the selected items in another way?

Thanks for the tip abot .toJSON method, the MST API is wonderul and sometimes I forgot all the methods 😄

**Comment by robinfehr** [comment]
@chemitaxis 
1. on your todo model, there is an afterCreate action
```js
.actions(self => ({ 
    // this is an action which gets triggered automatically after the store instance creation
    afterCreate() {
    }
})
```

about the progress: yes #504 will also cover that (does not yet)

about the workaround:
now you're using the onPatch within the afterCreate action mentioned above.
you can do this instead
```js
const store = Store.create({});
onPatch(store, patch => {
    console.dir("Got change: ", patch)
})
```

2. 
the benefit would be that you don't have to store the todos twice and therefore won't be within the json when you serialize (use toJSON) the store. 

the code to make that happen would look like
```js
const Todo = types
  .model("Todo", {
    id: types.identifier(),
    title: types.string,
    done: false,
    // add this instead
    selected: false
  }).actions(self => ({
    select(){
       self.selected = true;
    }
    unselect() {
      self.selected = false;
    }
    // ...
  })

...js
const Store = types
  .model("Store", {
    todos: types.optional(types.map(Todo), {}),
    history: types.optional(UndoManager, {}),
    // not needed anymore
    //selectedTodos: types.optional(types.array(types.reference(Todo)), [])
  }).views(self => ({
    get selectedTodos() {
       return self.todos.filter(todo => todo.isSelected);
    }
  })
```

hope it helps :)

**Comment by chemitaxis** [comment]
Hi @robinfehr!! Thanks for your answers... 

I have researched deeper, and I know why exactly the `store.history.canUndo` return `true`.

```
import { types, onPatch, getSnapshot } from "mobx-state-tree";
import { UndoManager } from "mst-middlewares";

const Todo = types.model("Todo", {
  id: types.identifier(),
  title: types.string,
  done: false
});

const Store = types.model("Store", {
  todos: types.optional(types.map(Todo), {}),
  history: types.optional(UndoManager, {}),
  selectedTodos: types.optional(types.array(types.reference(Todo)), [])
});

// create an instance from a snapshot
const store = Store.create({});

console.log(store.history.canUndo); // output: true, why? I just created the store...

console.log(store.history.history);
```

The "problem" is with this line:

`selectedTodos: types.optional(types.array(types.reference(Todo)), [])` 

Like the selectedTodos is an array of optional todos, but has a default value, the default value creation is marked as an undo action. You can play with it here: https://codesandbox.io/s/q8k4wzp836

About the second point, we prefer to get the patch clean, because in our case, `selected` is not a property of the object in the database, and we need a correlation one to one from our Model in MST and our Model in the database to automatically apply patchs. In other words, in our use case, selected is just a property of the UI and not a data stored value.

I don't know how can I achieve this "logic", because I would not like modify my "concept" model in MST. Maybe @mweststrate or @mattiamanzati could bring more light into the use case. I think the solution could be play with `types.maybe` and refactoring the code... but I'm not sure...

Best!!

**Comment by robinfehr** [comment]
@chemitaxis yeah that's another pitfall - 
```
const Store = types.model("Store", {
  todos: types.optional(types.map(Todo), {}),
  selectedTodos: types.optional(types.array(types.reference(Todo)), []),
  history: types.optional(UndoManager, {}), // add your history at the end and it should work.
});
```

**Comment by chemitaxis** [comment]
Ok, changed the order solved my problem... I will wait to your PR and I
will make another PR to update the doc ;) thanks!

**Comment by mweststrate** [comment]
@robinfehr if I am correct, if `UndoManager` uses `afterAttach` (but only if it has a parent), in combination with https://github.com/mobxjs/mobx-state-tree/pull/544/, that should avoid the issue of the member order. I think.

**Comment by mweststrate** [comment]
@chemitaxis the update undo manager has been released as 1.3.0. Thanks a lot @robinfehr !

**Comment by chemitaxis** [comment]
I’m testing!! Thanks guys!!
